src/lib.rs

```diff
@@ -466,3 +483,443 @@ where
             .merge(transactional_bonsai_storage.tries.db())
     }
 }
+
+/// An alternative to [BonsaiStorage] that does not store data in a trie.
+///
+/// Use this to store data that does not need to be verified but still has to be revertible.
+pub struct RevertibleStorage<ChangeID, DB>
+where
+    DB: BonsaiDatabase,
+    ChangeID: id::Id,
+{
+    db: KeyValueDB<DB, ChangeID>,
+    cache_storage_modified: HashMap<TrieKey, InsertOrRemove<Vec<u8>>>,
```

Comment:
You don't need any temporary storage in this structure `KeyValueDB` already do it for you and provide the features to store changes etc. You should called its `insert` `remove` etc in your function and just call commit on it in the commit

Reply:
Having checked out the code, doesn't `KeyValue` also write to the db during insertion if a `batch` isn't provided? See [here](https://github.com/Trantorian1/bonsai-trie/blob/0aa818543ac53bdb23461253cc114428743bd49b/src/key_value_db.rs#L192).